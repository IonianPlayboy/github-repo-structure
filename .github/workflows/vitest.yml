name: Vitests Tests

on: push

jobs:
    test:
        runs-on: ubuntu-latest
        steps:
            - name: Check out code
              uses: actions/checkout@v2
            - name: Vitest run tests
              run: yarn && yarn test:unit:ci
