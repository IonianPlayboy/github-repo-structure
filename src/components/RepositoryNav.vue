<template>
	<h3
		class="flex flex-wrap items-end whitespace-nowrap border-b border-slate-700 px-4 py-5 text-xl font-medium leading-6 text-gray-100 sm:px-6"
		data-testid="repositoryNav"
	>
		<RouterLink :to="baseRoute" data-testid="repository-root">
			<span class="text-primary">
				{{ owner }}
			</span>
			<span class="mx-1 text-primary-dark"> / </span>
			<span
				:class="{
					'animate-gradient from-teal-300 via-indigo-500 to-primary-lighter':
						!nodes.length,
					'from-primary-light to-primary-dark': nodes.length,
				}"
				class="bg-gradient-to-r bg-[length:400%_400%] bg-clip-text text-transparent"
			>
				{{ repo }}
			</span>
		</RouterLink>
		<template v-for="(item, i) in nodes" :key="`node_${item}`">
			<ChevronRightIcon class="h-5 w-5 text-primary" />
			<RouterLink
				:to="getRouteForIndex(i)"
				:class="{
					'animate-gradient from-teal-300 via-indigo-500 to-primary-lighter':
						i === nodes.length - 1,
					'from-primary-light to-primary-dark':
						i !== nodes.length - 1,
				}"
				class="bg-gradient-to-r bg-[length:400%_400%] bg-clip-text text-transparent"
				:data-testid="`repository-${item}`"
			>
				{{ item }}
			</RouterLink>
		</template>
	</h3>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";
import { useRepositoryStore } from "@/stores/repository";

import { RouterLink } from "vue-router";
import { ChevronRightIcon } from "@heroicons/vue/outline";
import { computed } from "vue";

const store = useRepositoryStore();
const { owner, repo, nodes } = storeToRefs(store);

const baseRoute = computed(() => `/repos/${owner.value}/${repo.value}`);

const getRouteForIndex = (index: number) =>
	`${baseRoute.value}/${nodes.value.slice(0, index + 1).join("/")}`;
</script>
