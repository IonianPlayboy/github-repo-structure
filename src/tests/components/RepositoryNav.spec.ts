import { describe, it, expect, fn, vi, beforeEach } from "vitest";

import { mount } from "@vue/test-utils";
import { createTestingPinia } from "@pinia/testing";
import RepositoryNav from "@/components/RepositoryNav.vue";

describe("RepositoryNav", () => {
	beforeEach(() => {
		vi.mock("vue-router", () => ({
			RouterLink: {
				template: "<a :href='to'><slot /></a>",
				props: ["to"],
			},
		}));
	});
	it("renders properly", () => {
		const testingPinia = createTestingPinia({
			initialState: {
				repository: {
					owner: "aa",
					repo: "aa",
				},
			},
			createSpy: fn,
		});
		const wrapper = mount(RepositoryNav, {
			global: {
				plugins: [testingPinia],
			},
		});

		expect(wrapper.text()).toContain("aa / aa");
		expect(wrapper.html()).toMatchSnapshot();
	});
	it("only shows one link when no path is set", () => {
		const testingPinia = createTestingPinia({
			initialState: {
				repository: {
					owner: "testOwner",
					repo: "testRepo",
				},
			},
			createSpy: fn,
		});
		const wrapper = mount(RepositoryNav, {
			global: {
				plugins: [testingPinia],
			},
		});

		const links = wrapper.findAll("a");
		expect(links).toHaveLength(1);
		const [rootLink] = links;

		expect(rootLink.text()).toContain("testOwner / testRepo");
		expect(rootLink.attributes()).toMatchObject({
			href: "/repos/testOwner/testRepo",
		});
		expect(wrapper.html()).toMatchSnapshot();
	});
	it("shows one link per node when a path is set", () => {
		const testingPiniaA = createTestingPinia({
			initialState: {
				repository: {
					owner: "testOwner",
					repo: "testRepo",
					path: "testPath",
				},
			},
			createSpy: fn,
		});
		const wrapperA = mount(RepositoryNav, {
			global: {
				plugins: [testingPiniaA],
			},
		});

		const linksA = wrapperA.findAll("a");
		expect(linksA).toHaveLength(2);
		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		const [_, nodeLink] = linksA;

		expect(nodeLink.text()).toContain("testPath");
		expect(nodeLink.attributes()).toMatchObject({
			href: "/repos/testOwner/testRepo/testPath",
		});
		expect(wrapperA.html()).toMatchSnapshot();

		const testingPiniaB = createTestingPinia({
			initialState: {
				repository: {
					owner: "testOwner",
					repo: "testRepo",
					path: "test/path",
				},
			},
			createSpy: fn,
		});
		const wrapperB = mount(RepositoryNav, {
			global: {
				plugins: [testingPiniaB],
			},
		});
		const links = wrapperB.findAll("a");
		expect(links).toHaveLength(3);
		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		const [__, firstLink, secondLink] = links;

		expect(firstLink.text()).toContain("test");
		expect(firstLink.attributes()).toMatchObject({
			href: "/repos/testOwner/testRepo/test",
		});
		expect(secondLink.text()).toContain("path");
		expect(secondLink.attributes()).toMatchObject({
			href: "/repos/testOwner/testRepo/test/path",
		});
		expect(wrapperB.html()).toMatchSnapshot();
	});
});
