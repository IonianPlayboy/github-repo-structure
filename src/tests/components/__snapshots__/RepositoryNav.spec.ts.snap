// Vitest Snapshot v1

exports[`RepositoryNav > only shows one link when no path is set 1`] = `"<h3 class=\\"flex flex-wrap items-end whitespace-nowrap border-b border-slate-700 px-4 py-5 text-xl font-medium leading-6 text-gray-100 sm:px-6\\" data-testid=\\"repositoryNav\\"><a href=\\"/repos/testOwner/testRepo\\" data-testid=\\"repository-root\\"><span class=\\"text-primary\\">testOwner</span><span class=\\"mx-1 text-primary-dark\\"> / </span><span class=\\"animate-gradient from-teal-300 via-indigo-500 to-primary-lighter bg-gradient-to-r bg-[length:400%_400%] bg-clip-text text-transparent\\">testRepo</span></a></h3>"`;

exports[`RepositoryNav > renders properly 1`] = `"<h3 class=\\"flex flex-wrap items-end whitespace-nowrap border-b border-slate-700 px-4 py-5 text-xl font-medium leading-6 text-gray-100 sm:px-6\\" data-testid=\\"repositoryNav\\"><a href=\\"/repos/aa/aa\\" data-testid=\\"repository-root\\"><span class=\\"text-primary\\">aa</span><span class=\\"mx-1 text-primary-dark\\"> / </span><span class=\\"animate-gradient from-teal-300 via-indigo-500 to-primary-lighter bg-gradient-to-r bg-[length:400%_400%] bg-clip-text text-transparent\\">aa</span></a></h3>"`;

exports[`RepositoryNav > shows one link per node when a path is set 1`] = `
"<h3 class=\\"flex flex-wrap items-end whitespace-nowrap border-b border-slate-700 px-4 py-5 text-xl font-medium leading-6 text-gray-100 sm:px-6\\" data-testid=\\"repositoryNav\\"><a href=\\"/repos/testOwner/testRepo\\" data-testid=\\"repository-root\\"><span class=\\"text-primary\\">testOwner</span><span class=\\"mx-1 text-primary-dark\\"> / </span><span class=\\"from-primary-light to-primary-dark bg-gradient-to-r bg-[length:400%_400%] bg-clip-text text-transparent\\">testRepo</span></a><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke-width=\\"2\\" stroke=\\"currentColor\\" aria-hidden=\\"true\\" class=\\"h-5 w-5 text-primary\\">
    <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" d=\\"M9 5l7 7-7 7\\"></path>
  </svg><a href=\\"/repos/testOwner/testRepo/testPath\\" class=\\"animate-gradient from-teal-300 via-indigo-500 to-primary-lighter bg-gradient-to-r bg-[length:400%_400%] bg-clip-text text-transparent\\" data-testid=\\"repository-testPath\\">testPath</a></h3>"
`;

exports[`RepositoryNav > shows one link per node when a path is set 2`] = `
"<h3 class=\\"flex flex-wrap items-end whitespace-nowrap border-b border-slate-700 px-4 py-5 text-xl font-medium leading-6 text-gray-100 sm:px-6\\" data-testid=\\"repositoryNav\\"><a href=\\"/repos/testOwner/testRepo\\" data-testid=\\"repository-root\\"><span class=\\"text-primary\\">testOwner</span><span class=\\"mx-1 text-primary-dark\\"> / </span><span class=\\"from-primary-light to-primary-dark bg-gradient-to-r bg-[length:400%_400%] bg-clip-text text-transparent\\">testRepo</span></a><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke-width=\\"2\\" stroke=\\"currentColor\\" aria-hidden=\\"true\\" class=\\"h-5 w-5 text-primary\\">
    <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" d=\\"M9 5l7 7-7 7\\"></path>
  </svg><a href=\\"/repos/testOwner/testRepo/test\\" class=\\"from-primary-light to-primary-dark bg-gradient-to-r bg-[length:400%_400%] bg-clip-text text-transparent\\" data-testid=\\"repository-test\\">test</a><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke-width=\\"2\\" stroke=\\"currentColor\\" aria-hidden=\\"true\\" class=\\"h-5 w-5 text-primary\\">
    <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" d=\\"M9 5l7 7-7 7\\"></path>
  </svg><a href=\\"/repos/testOwner/testRepo/test/path\\" class=\\"animate-gradient from-teal-300 via-indigo-500 to-primary-lighter bg-gradient-to-r bg-[length:400%_400%] bg-clip-text text-transparent\\" data-testid=\\"repository-path\\">path</a></h3>"
`;
